modules/relayers/src/payment_adapter.rs

```diff
@@ -61,47 +59,106 @@ where
 				T::BridgedChain::ID,
 				RewardsAccountOwner::BridgedChain,
 			),
-			DeliveryReward::get(),
 		);
 
 		rewarded_relayers as _
 	}
 }
 
+impl<T, MI> DeliveryPayments<T::AccountId> for DeliveryConfirmationPaymentsAdapter<T, MI>
+where
+	T: Config + pallet_bridge_messages::Config<MI>,
+	MI: 'static,
+{
+	type Error = &'static str;
+
+	fn pay_reward(
+		lane_id: LaneId,
+		relayer: T::AccountId,
+		_total_messages: MessageNonce,
+		valid_messages: MessageNonce,
+		_actual_weight: Weight,
+	) {
+		if valid_messages == 0 {
+			return
+		}
+
+		let _ = LaneRelayers::<T>::try_mutate(lane_id, |maybe_lane_relayers| {
+			if let Some(lane_relayers) = maybe_lane_relayers {
+				// if relayer is NOT in the active set, we don't want to do anything here
+				let relayer_in_active_set = lane_relayers
+					.active_relayers()
+					.iter()
+					.filter(|r| *r.relayer() == relayer)
+					.next()
+					.cloned();
+				let relayer_in_active_set = match relayer_in_active_set {
+					Some(relayer_in_active_set) => relayer_in_active_set,
+					None => return Err(()),
+				};
+
+				// if relayer is already in the active set, we don't want to do anything here
+				let is_in_next_set = lane_relayers
+					.next_relayers()
+					.iter()
+					.filter(|r| *r.relayer() == relayer)
+					.next()
+					.is_some();
+				if is_in_next_set {
+					return Err(())
+				}
+
+				// if relayer is not willoing to work on that lane anymore, we don't want to do
+				// anything here
+				let wants_to_work_on_lane = Pallet::<T>::registered_relayer(&relayer)
+					.map(|registration| registration.lanes().contains(&lane_id))
+					.unwrap_or(false);
+				if wants_to_work_on_lane {
+					return Err(())
+				}
+
+				if !lane_relayers.next_set_try_push(relayer, *relayer_in_active_set.reward()) {
+					return Err(())
+				}
+			}
+
+			Ok(())
+		});
+	}
+}
+
 // Update rewards to given relayers, optionally rewarding confirmation relayer.
 fn register_relayers_rewards<T: Config>(
-	confirmation_relayer: &T::AccountId,
+	_confirmation_relayer: &T::AccountId,
 	relayers_rewards: RelayersRewards<T::AccountId>,
-	lane_id: RewardsAccountParams,
-	delivery_fee: T::Reward,
+	reward_account: RewardsAccountParams,
 ) {
-	// reward every relayer except `confirmation_relayer`
-	let mut confirmation_relayer_reward = T::Reward::zero();
 	for (relayer, messages) in relayers_rewards {
-		// sane runtime configurations guarantee that the number of messages will be below
-		// `u32::MAX`
-		let relayer_reward = T::Reward::saturated_from(messages).saturating_mul(delivery_fee);
-
-		if relayer != *confirmation_relayer {
-			Pallet::<T>::register_relayer_reward(lane_id, &relayer, relayer_reward);
-		} else {
-			confirmation_relayer_reward =
-				confirmation_relayer_reward.saturating_add(relayer_reward);
-		}
+		// TODO: THIS IS WRONG - `LaneRelayers` is for relayers, which are delivering messages from BRIDGED to THIS
```

Comment:
I've missed that - the extension that boost priorities works at target chain. And reward is currently paid (`register_relayers_rewards` is called) at the source chain. So we do not know the reward that the relayer works for here (at the source chain). We will probably need to change the unrewarded relayers set to also store accumulated reward OR something else. Need to think about it a bit

Reply:
Ok, so it has three solutions imo:
1) to pay reward at the target chain (target BH). I'm not currently considering it, but it can be implemented later by slightly changing the solution (3);
2) to include proof of `LaneRelayers` into confirmation transaction. This has some obvious flaws - e.g. when the set changes, while there are unconfirmed messages, we may need to provide proof of old set and of the new set. Also we'll need to split delivered messages into two (or even more groups) - one, delivered by the old set and one, delivered by the new set. Also, it'll make confirmation transaction heavier, which is not good;
3) let's then just compute total accumulated relayer reward into `InboundLaneData::relayers` vector. So we'll "fix" the reward during delivery transaction and then will simply report it back to the source chain, where it'll actually be paid. We could later implement (1) or implenent it partially by registering lower reward (or zero if reward is paid fully at the target chain). To be super-careful, we may limit the maximal paid reward at the source chain by `DeliveredMessages * MaxRewardPerMessage`. The `MaxRewardPerMessage` shall be represented on both sides of the bridge, though.

That said, I'm gonna implement this in this PR, but then will extract it into a separate PR (after finishing implementation). This PR is already non-trivial.

Reply:
Changed in https://github.com/paritytech/parity-bridges-common/pull/2605/commits/5df0872b9ab3c9a02276fa6987be3fe0c195be8d. Need more tests + self review

Reply:
Two other relevant commits: https://github.com/paritytech/parity-bridges-common/pull/2605/commits/1a3efe764bf8f62c6912252bf16402d063849595 + https://github.com/paritytech/parity-bridges-common/pull/2605/commits/9c134da66aef091f2c8f5ed0081e3eb8e9466844 . I had to fix `RewardAtSource` type to be `u64`. We can't make it generic, because relayers pallet works with all chains, which may use different types for rewards (`MultiAssets`, `u128`, ...). At the same time, we need to cmp them in the pallet. So it is fixed to `u64`, but the other (chain) source may map it to its tokens using any kind of conversions.